### PR TITLE
Update protobuf version to 3.8.0

### PR DESCRIPTION
**Pull request by haon4**

### CODE REVIEW

**Comment by jtattermusch** [comment]
I'm seeing a failing distribtest:
https://source.cloud.google.com/results/invocations/5d258356-c24f-4f4f-bf4c-92f14cd29cac/targets/github%2Fgrpc/tests
```
+ cd examples/cpp/helloworld
+ make
protoc -I ../../protos --cpp_out=. ../../protos/helloworld.proto
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o helloworld.pb.o helloworld.pb.cc
protoc -I ../../protos --grpc_out=. --plugin=protoc-gen-grpc=`which grpc_cpp_plugin` ../../protos/helloworld.proto
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o helloworld.grpc.pb.o helloworld.grpc.pb.cc
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_client.o greeter_client.cc
g++ helloworld.pb.o helloworld.grpc.pb.o greeter_client.o -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc` -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_client
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_server.o greeter_server.cc
g++ helloworld.pb.o helloworld.grpc.pb.o greeter_server.o -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc` -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_server
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_async_client.o greeter_async_client.cc
g++ helloworld.pb.o helloworld.grpc.pb.o greeter_async_client.o -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc` -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_async_client
g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_async_client2.o greeter_async_client2.cc
g++ helloworld.pb.o helloworld.grpc.pb.o greeter_async_client2.o -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc` -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_async_client2
/usr/bin/ld: greeter_async_client2.o: undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make: *** [greeter_async_client2] Error 1
Makefile:51: recipe for target 'greeter_async_client2' failed
```

File: templates/grpc.gemspec.template [hunk 1]
```diff
@@ -31,7 +31,7 @@
     s.require_paths = %w( src/ruby/lib src/ruby/bin src/ruby/pb )
     s.platform      = Gem::Platform::RUBY
 
-    s.add_dependency 'google-protobuf', '~> 3.7'
+    s.add_dependency 'google-protobuf', '~> 3.8.0'
```
Comment (jtattermusch) [inline templates/grpc.gemspec.template]: I believe just `3.8` is correct for ruby. @apolcyn will know more.

File: bazel/grpc_deps.bzl [hunk 2]
```diff
@@ -115,9 +115,9 @@ def grpc_deps():
             url = "https://boringssl.googlesource.com/boringssl/+archive/afc30d43eef92979b05776ec0963c9cede5fb80f.tar.gz",
         )
 
-    if "com_github_madler_zlib" not in native.existing_rules():
```
Comment (jtattermusch) [inline bazel/grpc_deps.bzl]: why exactly is this change needed? It looks like you're only renaming the library. Is that because we need to use the same name that protobuf uses?

**Comment by jtattermusch** [comment]
This looks mostly good, but there's a few more things to fix (see other comments).

File: bazel/grpc_deps.bzl (hunk 2 above)
Comment (haon4) [reply bazel/grpc_deps.bzl]: Yea. Because protobuf somehow expects @zlib//:zlib to be present, and I don't want to build zlib twice: https://github.com/protocolbuffers/protobuf/blob/master/BUILD#L22

File: templates/grpc.gemspec.template (hunk 1 above)
Comment (haon4) [reply templates/grpc.gemspec.template]: Done. But I do see 3.8.0 here: https://rubygems.org/gems/google-protobuf, and it seems to work fine. Maybe 3.8 would automatically catch 3.8.0.1 or 3.8.1 release. Would we want to make changes to GRPC when that happen or just let it upgrade automatically?

**Comment by haon4** [comment]
> I'm seeing a failing distribtest:
> https://source.cloud.google.com/results/invocations/5d258356-c24f-4f4f-bf4c-92f14cd29cac/targets/github%2Fgrpc/tests
> 
> ```
> + cd examples/cpp/helloworld
> + make
> protoc -I ../../protos --cpp_out=. ../../protos/helloworld.proto
> g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o helloworld.pb.o helloworld.pb.cc
> protoc -I ../../protos --grpc_out=. --plugin=protoc-gen-grpc=`which grpc_cpp_plugin` ../../protos/helloworld.proto
> g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o helloworld.grpc.pb.o helloworld.grpc.pb.cc
> g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_client.o greeter_client.cc
> g++ helloworld.pb.o helloworld.grpc.pb.o greeter_client.o -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc` -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_client
> g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_server.o greeter_server.cc
> g++ helloworld.pb.o helloworld.grpc.pb.o greeter_server.o -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc` -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_server
> g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_async_client.o greeter_async_client.cc
> g++ helloworld.pb.o helloworld.grpc.pb.o greeter_async_client.o -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc` -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_async_client
> g++ -std=c++11 `pkg-config --cflags protobuf grpc`  -c -o greeter_async_client2.o greeter_async_client2.cc
> g++ helloworld.pb.o helloworld.grpc.pb.o greeter_async_client2.o -L/usr/local/lib `pkg-config --libs protobuf grpc++ grpc` -Wl,--no-as-needed -lgrpc++_reflection -Wl,--as-needed -ldl -o greeter_async_client2
> /usr/bin/ld: greeter_async_client2.o: undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'
> //lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
> collect2: error: ld returned 1 exit status
> make: *** [greeter_async_client2] Error 1
> Makefile:51: recipe for target 'greeter_async_client2' failed
> ```

Fixed.

**Comment by haon4** [comment]
Ping?

**Comment by jtattermusch** [comment]
Known failures:
https://github.com/grpc/grpc/issues/18416
https://github.com/grpc/grpc/issues/18608
